### PR TITLE
sort scenario selector

### LINKED ISSUE #1301: Scenario selector - Sort scenarios
Here's an example: I am doing a what-if analysis on a neural network by changing the parameter alpha. When I create a new scenario, "alpha 2.00", it gets put at the top of the list instead of being ordered alphabetically. The end user should be able to sort scenarios in the scenario_selector.

core_issues_1_1

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1301

image

### CODE REVIEW

**Review by namnguyen20999 — DISMISSED** [review]
LGTM

File: frontend/taipy-gui/src/components/Taipy/TableSort.spec.tsx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import userEvent from "@testing-library/user-event";
+
+import TableSort from "./TableSort";
+import { ColumnDesc } from "./tableUtils";
+
+const tableColumns = {
+    StringCol: { dfid: "StringCol", type: "object", index: 0, format: "", filter: true },
+    NumberCol: { dfid: "NumberCol", type: "int", index: 1, format: "", filter: true },
+    BoolCol: { dfid: "BoolCol", type: "bool", index: 2, format: "", filter: true },
+    DateCol: { dfid: "DateCol", type: "datetime", index: 2, format: "", filter: true },
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.spec.tsx]:
```suggestion
    DateCol: { dfid: "DateCol", type: "datetime", index: 3, format: "", filter: true },
```
(not really useful though)

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+const orderCaptionSx = { ml: 1 };
+
+const getSortDesc = (columns: Record<string, ColumnDesc>, colId?: string, asc?: boolean) => {
+    if (colId && asc !== undefined) {
+        return {
+            col: columns[colId].dfid,
+            order: !!asc,
+        } as SortDesc;
+    }
+};
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.tsx]: No 'else' statement?

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        const fd = getSortDesc(columns, colId, order);
+        fd && setSort(idx, fd);
+    }, [idx, setSort, columns, colId, order]);
+
+    useEffect(() => {
+        if (sort && idx > -1) {
+            const col = Object.keys(columns).find((col) => columns[col].dfid === sort.col) || "";
+            setColId(col);
+            setOrder(sort.order);
+            setEnableCheck(false);
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.tsx:121]: `false`?

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                                {columns[col].title || columns[col].dfid}
+                            </MenuItem>
+                        ))}
+                    </Select>
+                </FormControl>
+            </Grid>
+            <Grid item xs={4}>
+                <FormControlLabel
+                    control={<Switch checked={order} onChange={onOrderSwitch} />}
+                    label="Order"
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.tsx]: I would not add a label for this one. 'asc'/'desc' look clear enough to me...

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                        newSds = [...sds];
+                    } else {
+                        newSds = sds.map((fd, index) => (index == idx ? nsd : fd));
+                    }
+                } else if (remove) {
+                    newSds = sds;
+                } else {
+                    newSds = [...sds, nsd];
+                }
+                onValidate([...newSds]);
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.tsx:205]:
```suggestion
                onValidate(newSds];
```
?

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+    useEffect(() => {
+        columns &&
+            appliedSorts &&
+            setSorts(appliedSorts.filter((fd) => Object.values(columns).some((cd) => cd.dfid === fd.col)));
+    }, [columns, appliedSorts]);
+
+    return (
+        <>
+            <Tooltip title={`${sorts.length} sort${sorts.length > 1 ? "s" : ""} applied`}>
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/TableSort.tsx:220]: Sure about the outcome??

File: frontend/taipy/src/CoreSelector.tsx [hunk 7]
```diff
@@ -528,7 +569,7 @@ const CoreSelector = (props: CoreSelectorProps) => {
     return (
         <>
             <Grid container sx={switchBoxSx} gap={1}>
-                {active && colFilters ? (
+            {active && colFilters ? (
                     <Grid item>
```
Comment (FabienLelaquais) [inline frontend/taipy/src/CoreSelector.tsx:573]: To me this block should be indented 4 chars back.

File: taipy/gui_core/_adapters.py [hunk 8]
```diff
@@ -264,7 +257,7 @@ def _get_datanode_property(attr: str):
 
 
 class _GuiCoreScenarioProperties(_TaipyBase):
-    __SC_TYPES = {
+    _SC_TYPES = {
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py:260]: I guess you don't need the "_SC" prefix: you're in the `_GuiCoreScenarioProperties` class already.
`_TYPES` is clear enough in my opinion.

File: taipy/gui_core/_adapters.py [hunk 9]
```diff
@@ -318,19 +320,57 @@ def get(self):
...
+                    if self.full_desc()
+                    else (attr,)
                     for attr in flist
                     if attr and isinstance(attr, str)
                 ]
             )
         return None
+
+
+class _GuiCoreScenarioFilter(_GuiCoreScenarioProperties):
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py:343]: 👍

File: taipy/gui_core/_context.py [hunk 10]
```diff
@@ -203,7 +205,7 @@ def submission_status_callback(self, submission_id: t.Optional[str]):
     def no_change_adapter(self, entity: t.List):
         return entity
 
-    def cycle_adapter(self, cycle: Cycle):
+    def cycle_adapter(self, cycle: Cycle, sorts: t.Optional[t.List[t.Dict[str, t.Any]]] = None):
```
Comment (FabienLelaquais) [inline taipy/gui_core/_context.py:208]: Not sure about the name for 'sorts', but nothing better to propose that would be short enough...

File: taipy/gui_core/_context.py [hunk 11]
```diff
@@ -257,8 +256,27 @@ def adapt_scenarios(self, cycle: t.List):
...
+                order = sd.get("order", True)
+                sorted_list = sorted(sorted_list, key=_GuiCoreContext.get_entity_property(col), reverse=not order)
+        else:
+            sorted_list = sorted(entities, key=_GuiCoreContext.get_entity_property("creation_date"))
+        return [self.cycle_adapter(e, sorts) if isinstance(e, Cycle) else e for e in sorted_list]
+
     def get_scenarios(
-        self, scenarios: t.Optional[t.List[t.Union[Cycle, Scenario]]], filters: t.Optional[t.List[t.Dict[str, t.Any]]]
+        self,
+        scenarios: t.Optional[t.List[t.Union[Cycle, Scenario]]],
```
Comment (FabienLelaquais) [inline taipy/gui_core/_context.py:277]: I like (just like in get_sorted_entity_list()) that you don't provide default values... 👍

File: taipy/gui_core/viselements.json [hunk 12]
```diff
@@ -100,8 +100,21 @@
                     },
                     {
                         "name": "filter",
-                        "type": "str|list[str]",
-                        "doc": "TODO: a list of scenario attributes to filter on."
+                        "type": "bool|str|list[str]",
+                        "default_value": "Config id;Label;Creation date;Cycle label;Cycle start;Cycle end;Primary;Tags",
```
Comment (FabienLelaquais) [inline taipy/gui_core/viselements.json:104]: Are all attributes listed here? And if yes should this really be the default?

**Review by FabienLelaquais — DISMISSED** [review]
👍really nice

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx (hunk 5 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/TableSort.tsx:205]: need to mutate the array

File: taipy/gui_core/_adapters.py (hunk 8 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/_adapters.py:260]: we also have DN_ for Datanode

File: taipy/gui_core/viselements.json (hunk 12 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/viselements.json:104]: from the spec doc

File: frontend/taipy-gui/src/components/Taipy/TableSort.tsx (hunk 3 above)
Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/TableSort.tsx:121]: yes, it will be changed to true if something changes

**Review by FabienLelaquais — DISMISSED** [review]
💪

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

**Review by dinhlongviolin1 — APPROVED** [review]
LGTM

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**